pkg/controllers/management/aks/aks_cluster_handler.go

```diff
@@ -53,23 +59,51 @@ func Register(ctx context.Context, wContext *wrangler.Context, mgmtCtx *config.M
 	}
 
 	aksCCDynamicClient := mgmtCtx.DynamicClient.Resource(aksClusterConfigResource)
-	e := &aksOperatorController{clusteroperator.OperatorController{
-		ClusterEnqueueAfter:  wContext.Mgmt.Cluster().EnqueueAfter,
-		SecretsCache:         wContext.Core.Secret().Cache(),
-		TemplateCache:        wContext.Mgmt.CatalogTemplate().Cache(),
-		ProjectCache:         wContext.Mgmt.Project().Cache(),
-		AppLister:            mgmtCtx.Project.Apps("").Controller().Lister(),
-		AppClient:            mgmtCtx.Project.Apps(""),
-		NsClient:             mgmtCtx.Core.Namespaces(""),
-		ClusterClient:        wContext.Mgmt.Cluster(),
-		CatalogManager:       mgmtCtx.CatalogManager,
-		SystemAccountManager: systemaccount.NewManager(mgmtCtx),
-		DynamicClient:        aksCCDynamicClient,
-		ClientDialer:         mgmtCtx.Dialer,
-		Discovery:            wContext.K8s.Discovery(),
-	}}
+	e := &aksOperatorController{
+		OperatorController: clusteroperator.OperatorController{
+			ClusterEnqueueAfter:  wContext.Mgmt.Cluster().EnqueueAfter,
+			SecretsCache:         wContext.Core.Secret().Cache(),
+			TemplateCache:        wContext.Mgmt.CatalogTemplate().Cache(),
+			ProjectCache:         wContext.Mgmt.Project().Cache(),
+			AppLister:            mgmtCtx.Project.Apps("").Controller().Lister(),
+			AppClient:            mgmtCtx.Project.Apps(""),
+			NsClient:             mgmtCtx.Core.Namespaces(""),
+			ClusterClient:        wContext.Mgmt.Cluster(),
+			CatalogManager:       mgmtCtx.CatalogManager,
+			SystemAccountManager: systemaccount.NewManager(mgmtCtx),
+			DynamicClient:        aksCCDynamicClient,
+			ClientDialer:         mgmtCtx.Dialer,
+			Discovery:            wContext.K8s.Discovery(),
+		},
+		secretClient: wContext.Core.Secret(),
+	}
 
 	wContext.Mgmt.Cluster().OnChange(ctx, "aks-operator-controller", e.onClusterChange)
+	wContext.Core.Secret().OnChange(ctx, "aks-cloud-credential", e.onCloudCredentialChange)
+}
+
+func (e *aksOperatorController) onCloudCredentialChange(key string, cred *corev1.Secret) (*corev1.Secret, error) {
+	// credential doesn't exist or is deleting
+	if cred == nil || cred.DeletionTimestamp != nil {
+		return cred, nil
+	}
+	// credential is not for azure
+	subscriptionID, ok := cred.Data["azurecredentialConfig-subscriptionId"]
+	if !ok {
+		return cred, nil
+	}
+	// credential already has tenant ID
+	tenantIDBytes, ok := cred.Data["azurecredentialConfig-tenantId"]
+	if ok && string(tenantIDBytes) != "" {
+		return cred, nil
+	}
+	tenantID, err := azureutil.FindTenantID(context.Background(), azure.PublicCloud, string(subscriptionID))
+	if err != nil {
+		return cred, err
+	}
+	cred = cred.DeepCopy()
+	cred.Data["azurecredentialConfig-tenantId"] = []byte(tenantID)
+	return e.secretClient.Update(cred)
```

Comment:
Would be curious if UI could create a race condition here. Like user creates it and tries to use it before tenant exists. Talking offline and we are assuming not.